### PR TITLE
A Hypothesis Test

### PR DESCRIPTION
**Pull request by Julian**
Hi :)

[Hypothesis](http://hypothesis.readthedocs.io/en/latest/) just in case you haven't heard of it, is a really great piece of software for doing QuickCheck-style invariant checking -- really useful I think in cases like HdrHistogram.

I have some hypothesis tests internally for some of our usage of HdrHistogram_Py, but I think if you're interested, adding some upstream would be useful.

As a first shot, I tried adding one to ensure the invariant that seems to be indicated in [a docstring](https://github.com/HdrHistogram/HdrHistogram_py/blob/master/hdrh/histogram.py#L52-L62) but I actually haven't been able to get it to pass :)

Possibly I've made an error, seems at least likely given how much time I've spent debugging, but was hoping to submit it and collaborate on getting it passing, whether it's a testing bug or not.

Lemme know if that sounds useful.

### CODE REVIEW

**Comment by Julian** [comment]
Seems like if I use `numpy.percentile(..., interpolation="lower")` I get a bit further success-wise, but still not all the way there, because `[1, 2, 2, 2, 2]` HdrHistogram claims has a 25 percentile of 1, but all interpolation modes in numpy consider that to be 2, which is 50% off obviously.

**Comment by Julian** [comment]
```
>>> numpy.percentile([1,2], 50.00000000000001, interpolation="nearest")
2

>>> h.get_value_at_percentile(50.00000000000001)
1
```

seems like another example, though there I suspect there possibly is an actual issue with some lost precision potentially?

**Comment by ahothan** [comment]
Hi Julian, Thanks for your suggestion using Hypothesis, I did not know about it but it looks like a terrific library and pretty intuitive to use! It can be a very good addition to any test code when used appropriately. I would like to explore adding its use to the main branch with your help.
Regarding the test that was added, there might be a misunderstanding on the semantic of the significant figure argument when creating a histogram that may explain why your test fails. The significant figure will tell HdrHistogram the required precision of the each value recorded across the allowed range. For example a precision of 2, means the histogram will alway keep recorded values within 0.01% of its actual value. For example, if the range is 1 usec to 100 seconds, the recorded value at the 1 usec level will be +/- 0.01usec and at the 100 second level will be +/- 1 sec.
This is directly derived from the way buckets are being carved for counting intervals. On a simple histogram, buckets are linear (meaning each bucket represents a fixed interval). This obviously does not scale well for large ranges and small precision. HdrHistogram implements non linear buckets where each interval is dependent on the precision level: 1 usec level will have intervals of 0.01 usec while 1 sec level will have intervals of 1 sec. This helps in cutting down the number of buckets while preserving a relative precision at every interval level - hence the name High Dynamic Range histogram. This stems from the observation that if you care about a high dynamic range (say 1 usec to 1 day) in general you want to be very accurate at the low range but it does not make usually sense to be accurate to the sub usec level when your numbers go as high a multiple hours (because at that level, you;d have much bigger fish to fry to chase sub usec values).

**Comment by ahothan** [comment]
Regarding the 2 examples, I'd need to look at them more in detail to see why they differ from numpy. There could be some rounding/precision that may explain differences, or those might be bugs as well.
Clearly the difference between what numpy returns and what hdr returns has nothing to do with the significant digit parameter as explained above. Those could also be explained by the design tradeoffs in the hdr code (which was mostly ported from the Java version), while numpy is likely a more precise implementation but likely not as scalable in constrained runtime environments.
So the test as written is incorrect anyway.

**Comment by Julian** [comment]
Thanks for the comments!

> I would like to explore adding its use to the main branch with your help.
Regarding the test that was added, there might be a misunderstanding on the semantic of the significant figure argument when creating a histogram that may explain why your test fails. [..snip..]

I think all of this makes sense to me, let me just verify then whether I've still made a mistake -- that means that the *relative* error then is what should always be less than ``10 ^ -(significant figures)`` right? Which is the intent of the assertion in the test:

https://github.com/HdrHistogram/HdrHistogram_py/pull/14/files#diff-4be43396b98bb31a1866316a8d5ab040R150

Basically just to speak it out for myself as well, the way that test works is:

* randomly generate a histogram
* randomly generate some values to record and record them
* randomly generate a percentile to test, and assert that the value at that percentile is within the guaranteed relative error.

> what would be useful is to run the same 2 examples with the Java and/or C version of hdr and see what they returm

Will try to give that a shot now, though my Java and C skills don't particularly rate very well.

**Comment by Julian** [comment]
So I *think* despite their shortcomings that I've managed it :) and it does look like the C version is giving the same answers. Code:

```c
#include <inttypes.h>
 
#include <hdr_histogram.h>

int main()
{
    struct hdr_histogram* histogram;
    hdr_init(1, INT64_C(3600000000), 3, &histogram);

    hdr_record_value(histogram, 1);
    hdr_record_value(histogram, 2);
    hdr_record_value(histogram, 2);
    hdr_record_value(histogram, 2);
    hdr_record_value(histogram, 2);

    int64_t value = hdr_value_at_percentile(histogram, 25);
    printf("%" PRId64 "\n", value);

    struct hdr_histogram* histogram2;
    hdr_init(1, INT64_C(3600000000), 3, &histogram2);

    hdr_record_value(histogram2, 1);
    hdr_record_value(histogram2, 2);

    int64_t value2 = hdr_value_at_percentile(histogram2, 50.00000000000001);
    printf("%" PRId64 "\n", value2);
}
```

which outputs:

```
1
1
```

The core question then sounds like it's just:

* "What's the 25th percentile of [1, 2, 2, 2, 2]"

and

* "What's the 50.00000000000001th percentile of [1, 2]"

right?

These are going to be the simplest examples hypothesis finds, but it will produce countless others if we want more examples of cases where the invariant is failing. Would love to get a sense for what the actual guarantee is then.

**Comment by Julian** [comment]
A little more digging here -- haven't gotten all the way to the bottom of it, but I think this also has something to do with the size of the data set? That small data sets have inaccurate percentile calculations -- specifically, lines like https://github.com/HdrHistogram/HdrHistogram_py/blob/1cfd30573a78aaa9626408cbaf752b8541bce3c8/hdrh/histogram.py#L264 seem a bit fishy?

**Comment by giltene** [comment]
It's not the significant figure accuracy. It's how the HdrHistogram getValueAtPercentile chose to draw the line in the counts. The [current] relevant math is driven by this [java] logic.

final double requestedPercentile = Math.min(percentile, 100.0); // Truncate down to 100%
long countAtPercentile = (long)(((requestedPercentile / 100.0) * getTotalCount()) + 0.5); // round to nearest
countAtPercentile = Math.max(countAtPercentile, 1);

The choice to round the count to the nearest count (as opposed to consistently truncate it or round it up) is the reasons that e.g. the 25%'ile of [1,2,2,2,2] is 1, and the 30%'ile of the same is 2.

There is [probably] a good argument that this is ambiguous and potentially inconsistent with the method's specification. However, since this is how HdrHistogram behaves across various language implementations, I don't want to jump and change this logic quickly. Lets discuss more widely on gitter.

**Comment by Julian** [comment]
@giltene hi! Glad to run into you :)

That Java line looks exactly like the line I found, so glad I was on the right track.

Is the Gitter public? I doubt I have much to contribute here but would love to watch the conversation.

**Comment by Julian** [comment]
Never mind, just found the link :)

**Comment by ahothan** [comment]
Closing this PR as there is no more activity on it

**Comment by Julian** [comment]
Does it pass on tip these days? If so it likely is ready to merge as-is, if you want it, no?

And if not, things likely still do not meet the guarantee in the readme.

**Comment by ahothan** [comment]
Sorry, I was under the wrong impression this PR was no longer active, I can merge it in

**Comment by ahothan** [comment]
can you please resolve conflicts, recheck the test is passing

**Comment by Julian** [comment]
No worries -- rebased, but looks like it still fails unfortunately.
